### ISSUE TITLE
Generics use case (partially covered)

### ISSUE THREAD

**Issue by camilledejoye** [phpactor/phpactor#1530]
Hi,

I'm really enjoying the new type system, great job again!

In the same line as the previous issue here is a small test scenario to show case some interesting use cases that could be added to the list of tests:
```php
<?php

namespace Test;

interface Foo {
    public function bar(): void;
}

interface Foos {
    /**
     * @return Collection<Foo>
     */
    public function all(): Collection;
}

/**
 * @template TKey of array-key
 * @template T
 * @template-extends \IteratorAggregate<TKey, T>
 * @template-extends \ArrayAccess<TKey|null, T>
 */
interface Collection extends \IteratorAggregate, \ArrayAccess
{
    /**
     * @return T
     */
    public function first();

    /**
     * @return Collection<TKey, T>
     */
    public function filter(\Closure $p);
}

function (Foos $foos) {
    foreach ($foos->all() as $foo) {
        $foo->bar(); // OK
    }

    $foos->all()->first()->bar(); // Not recognized
    $foos->all()[0]->bar(); // Not recognized

    $filtered = $foos->all()->filter(fn($foo) => $foo);
    foreach ($filtered as $foo) {
        $foo->bar(); // Not recognized
    }
}
```

**Comment by dantleech** [comment]
yeah generics isn't 100% yet, i'll use this as a test case when I get back to it, thanks :)

**Comment by dantleech** [comment]
Fixed with the exception of the ArrayAccess (#1593 ) in #1595 (see test case). In addition there is a follow up issue for further support in this direction #1594
